state-indexer/src/main.rs

```diff
@@ -23,15 +23,38 @@ pub(crate) const INDEXER: &str = "state_indexer";
 async fn handle_streamer_message(
     streamer_message: near_indexer_primitives::StreamerMessage,
     db_manager: &(impl database::StateIndexerDbManager + Sync + Send + 'static),
+    rpc_client: &near_jsonrpc_client::JsonRpcClient,
     indexer_id: &str,
     stats: std::sync::Arc<tokio::sync::RwLock<metrics::Stats>>,
 ) -> anyhow::Result<()> {
     let block_height = streamer_message.block.header.height;
     let block_hash = streamer_message.block.header.hash;
+
+    let new_epoch_id = streamer_message.block.header.epoch_id;
+
+    // handle first indexing epoch
+    let mut stats_lock = stats.write().await;
+    let current_epoch_id = if let Some(current_epoch_id) = stats_lock.current_epoch_id {
+        current_epoch_id
+    } else {
+        let (epoch_id, height) =
+            handle_epoch(new_epoch_id, None, stats_lock.current_epoch_height, rpc_client, db_manager).await?;
+        stats_lock.current_epoch_id = Some(epoch_id);
+        stats_lock.current_epoch_height = height;
+        epoch_id
+    };
+    drop(stats_lock);
+
     tracing::debug!(target: INDEXER, "Block height {}", block_height,);
```

Comment:
I'd like to have this line above any "heavy" logic. Please move it right below the definition of the `new_epoch_id`